### PR TITLE
dockerfile: Add a basic reference Dockerfile

### PR DESCRIPTION
**Pull request by sillyfrog**
This is a sample of how to get Klipper running in Docker, because of the
virtual serial port, the user will have to modify things to include
something to send the GCode. I have included a link to my repo which has
OctoPrint up and running as an example (please let me know if this is an issue).

Fixes: #358 (maybe)

### CODE REVIEW

**Comment by KevinOConnor** [comment]
Thanks.  Can the dockerfile be moved to the scripts directory, or is it dependent on being in the root of the klipper repo?

-Kevin

**Comment by sillyfrog** [comment]
Unfortunately the `Dockerfile` needs to be in the root of the repo, as Docker will only allow access to files at the same level and deeper when building an image.

**Comment by KevinOConnor** [comment]
Okay, thanks.  I missed that in the last commit.  I think we should move it to the scripts/ directory and update the comments to instruct the user to copy it to the main directory then.  (I'd prefer to limit the number of files in the top Klipper directory.)  I'll try to take a look at this in the next few days.

-Kevin

**Comment by sillyfrog** [comment]
Makes sense, what's there is very much a reference and people will likely need to edit it anyway, so asking them to move it is probably not a big deal (depending on how they do their version/change tracking etc).

**Comment by mamoit** [comment]
If you allow `sudo` to the user `klipper` without password and for all commands, why not run it as root?

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+### Klipper setup ###
+WORKDIR /home/klippy
+
+COPY . klipper/
+USER root
+RUN echo 'klippy ALL=(ALL:ALL) NOPASSWD: ALL' > /etc/sudoers.d/klippy && \
+    chown klippy:klippy -R klipper
+USER klippy
+RUN ls -lha ./klipper/scripts/
```
Comment (mamoit) [inline Dockerfile]: what is the point of this `ls`?

**Comment by mamoit** [comment]
Ok, got it.
The install script depends on sudo.

File: Dockerfile (hunk 1 above)
Comment (sillyfrog) [reply Dockerfile]: Ahhh... that would be left over debugging I missed. Fixed now :)

**Comment by KevinOConnor** [comment]
While you're making changes, it would be great if you move the file to the scripts/ directory and update the comments to indicate it should be copied to the main klipper directory before running.  Happy to commit it then.

-Kevin

**Comment by KevinOConnor** [comment]
Thanks.

**Comment by ddinternetagentur** [comment]
I need help with Docker installation.. I have created a Image with this lines but i have errors on startup..

~~~
FROM octoprint/octoprint

RUN /opt/octoprint/venv/bin/python -m pip install https://github.com/mmone/OctoPrintKlipper/archive/master.zip

USER root

RUN apt-get update && \
    apt-get install -y sudo

USER octoprint

WORKDIR /home/octoprint

RUN git clone https://github.com/KevinOConnor/klipper

USER root

RUN echo 'octoprint ALL=(ALL:ALL) NOPASSWD: ALL' > /etc/sudoers.d/octoprint

USER octoprint
RUN ./klipper/scripts/install-octopi.sh
RUN cp klipper/config/printer-anet-a8-2017.cfg /home/octoprint/printer.cfg

CMD ["/home/octoprint/klippy-env/bin/python", "/home/octoprint/klipper/klippy/klippy.py", "/home/octoprint/printer.cfg"]
~~~

ERROR:

~~~
INFO:root:Starting serial connect
ERROR:root:process_identify error
Traceback (most recent call last):
  File "/home/octoprint/klipper/klippy/msgproto.py", line 374, in process_identify
    all_messages = dict(commands)
TypeError: cannot convert dictionary update sequence element #0 to a sequence
ERROR:root:Protocol error
Traceback (most recent call last):
  File "/home/octoprint/klipper/klippy/klippy.py", line 137, in _connect
    cb()
  File "/home/octoprint/klipper/klippy/mcu.py", line 609, in _connect
    self._serial.connect()
  File "/home/octoprint/klipper/klippy/serialhdl.py", line 83, in connect
    msgparser.process_identify(identify_data)
  File "/home/octoprint/klipper/klippy/msgproto.py", line 386, in process_identify
    raise error("Error during identify: %s" % (str(e),))
error: Error during identify: cannot convert dictionary update sequence element #0 to a sequence
~~~

**Comment by sillyfrog** [comment]
@ddinternetagentur I'm not sure what that error is, but I would make sure the firmware is up to date. I would also suggest testing with the Dockerfiles at https://github.com/sillyfrog/OctoPrint-Klipper-mjpg-Dockerfile as this is what I use (so I know it works). You can then also file an issue there if it's still not working as expected.

**Pull request merged**